### PR TITLE
Zero-shot AutoML

### PR DESCRIPTION
**Pull request by sonichi**

### CODE REVIEW

File: website/docs/Use-Cases/Zero-Shot-AutoML.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+```python
+import sklearn.ensemble as ensemble
+from flaml.default import flamlize_estimator
+ExtraTreesClassifier = flamlize_estimator(
+    ensemble.ExtraTreesClassifier, "extra_tree", "classification"
+)
+```
+
+Then, you can share this "flamlized" `ExtraTreesClassifier` together with the location of your learned defaults with others (or the _future_ yourself). They will benefit from your past experience. Your group can also share experiences in a central place and update the learned defaults continuously. Overtime, your organization gets better collectively.
```
Comment (qingyun-wu) [inline website/docs/Use-Cases/Zero-Shot-AutoML.md:247]:
```suggestion
Then, you can share this "flamlized" `ExtraTreesClassifier` together with the location of your learned defaults with others (or the _future_ yourself). They will benefit from your past experience. Your group can also share experiences in a central place and update the learned defaults continuously. Over time, your organization gets better collectively.
```

File: website/docs/Use-Cases/Zero-Shot-AutoML.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+Read the next section to understand how to generate these files if you would like to meta-learn the defaults yourself.
+
+## How to Prepare Offline
+
+This section is intended for:
+1. AutoML providers for a particular domain.
+1. Data scientists or engineers who need to repeatedly train models for similar tasks with varying training data.
+
+Instead of running full hyperparmaeter tuning from scratch every time, one can leverage the tuning experiences in similar tasks before. While we have offered the meta-learned defaults from tuning experiences of several popular learners on benchmark datasets for classification and regression, you can customize the defaults for your own tasks/learners/metrics based on your own tuning experiences.
```
Comment (qingyun-wu) [inline website/docs/Use-Cases/Zero-Shot-AutoML.md:137]:
```suggestion
Instead of running full hyperparameter tuning from scratch every time, one can leverage the tuning experiences in similar tasks before. While we have offered the meta-learned defaults from tuning experiences of several popular learners on benchmark datasets for classification and regression, you can customize the defaults for your own tasks/learners/metrics based on your own tuning experiences.
```

File: website/docs/Getting-Started.md [hunk 3]
```diff
@@ -76,6 +76,16 @@ analysis = tune.run(
...
 
+#### [Zero-shot AutoML](Use-Cases/Zero-Shot-AutoML)
+
+FLAML offers a unique, seamless and effortless way to leverage AutoML for the commonly used classifiers and regressors such as LightGBM and XGBoost. For example, if you are using `lightgbm.LGBMClassifier` as your current learner, all you need to do is to replace `from ligthgbm import LGBMClassifier` by:
+
+```python
+from flaml.default import LGBMClassifier
+```
+
+Then, you can use it just like you use the original `LGMBClassifier`. Your other code can remain unchanged. When you call the `fit()` function from `flaml.default.LGBMClassifier`, it will automatically instantiates a good data-dependent hyperparameter configuration for your dataset, which is expected to work better than the default configuration.
```
Comment (qingyun-wu) [inline website/docs/Getting-Started.md:87]:
```suggestion
Then, you can use it just like you use the original `LGMBClassifier`. Your other code can remain unchanged. When you call the `fit()` function from `flaml.default.LGBMClassifier`, it will automatically instantiate a good data-dependent hyperparameter configuration for your dataset, which is expected to work better than the default configuration.
```

File: flaml/default/README.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+Read the next subsection to understand how to generate these files if you would like to meta-learn the defaults yourself.
+
+To perform hyperparameter search starting with the data-dependent defaults, remove `max_iter=0`.
+
+## Perform Meta Learning
+
+FLAML provides a package `flaml.default` to learn defaults customized for your own tasks/learners/metrics.
+
+### Prepare a collectioin of training tasks
```
Comment (qingyun-wu) [inline flaml/default/README.md:78]:
```suggestion
### Prepare a collection of training tasks
```

File: website/docs/Use-Cases/Zero-Shot-AutoML.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+### Prepare a collectioin of training tasks
```
Comment (qingyun-wu) [inline website/docs/Use-Cases/Zero-Shot-AutoML.md:139]:
```suggestion
### Prepare a collection of training tasks
```

File: flaml/default/README.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+adult,0,binary,0.919499,{'_modeljson': 'lgbm/connect-4.json'}
+adult,0,binary,0.930109,{'_modeljson': 'lgbm/default.json'}
+adult,0,binary,0.932453,{'_modeljson': 'lgbm/dilbert.json'}
+adult,0,binary,0.921959,{'_modeljson': 'lgbm/Dionis.json'}
+adult,0,binary,0.910763,{'_modeljson': 'lgbm/poker.json'}
+...
+```
+
+The `type` column indicates the type of the task, such as regression, binary or multiclass.
+The `result` column stores the evaluation result, assumed the large the better. The `params` column indicates which json config is used. For example 'lgbm/2dplanes.json' indicates that the best lgbm configuration extracted from 2dplanes is used.
```
Comment (qingyun-wu) [inline flaml/default/README.md:146]:
```suggestion
The `result` column stores the evaluation result, assuming the large the better. The `params` column indicates which json config is used. For example 'lgbm/2dplanes.json' indicates that the best lgbm configuration extracted from 2dplanes is used.
```

File: flaml/default/suggest.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+            - "version": a str of the version number.
+            - "preprocessing": a dictionary containing:
+                * "center": a list of meta feature value offsets for normalization.
+                * "scale": a list of meta feature scales to normalize each dimension.
+            - "neighbors": a list of dictionaries. Each dictionary contains:
+                * "features": a list of the normalized meta features for a neighbor.
+                * "choice": a integer of the configuration id in the portfolio.
+            - "portfolio": a list of dictionaries, each corresponding to a configuration:
+                * "class": a str of the learner name.
+                * "hyperparameters": a dict of the config. They key "FLAML_sample_size" will be ignored.
```
Comment (qingyun-wu) [inline flaml/default/suggest.py:115]:
```suggestion
                * "hyperparameters": a dict of the config. The key "FLAML_sample_size" will be ignored.

```

File: flaml/default/suggest.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        y: A series of labels in shape n*1.
+        estimator_or_predictor: A str of the learner name or a dict of the learned config predictor.
+            If a dict, it contains:
+            - "version": a str of the version number.
+            - "preprocessing": a dictionary containing:
+                * "center": a list of meta feature value offsets for normalization.
+                * "scale": a list of meta feature scales to normalize each dimension.
+            - "neighbors": a list of dictionaries. Each dictionary contains:
+                * "features": a list of the normalized meta features for a neighbor.
+                * "choice": a integer of the configuration id in the portfolio.
```
Comment (qingyun-wu) [inline flaml/default/suggest.py:112]:
```suggestion
                * "choice": an integer of the configuration id in the portfolio.

```

File: flaml/default/portfolio.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+
+    Args:
+        meta_features: metafeatures matrix
+        regret: regret matrix
+        strat: one of {greedy-feedback, greedy}
+    """
+    if strat == "greedy-feedback":
+        portfolio = greedy.smart(regret, meta_features, regret_bound)
+    elif strat == "greedy":
+        portfolio = greedy.smart(regret, None, regret_bound)
```
Comment (qingyun-wu) [inline flaml/default/portfolio.py:53]: What if strat is neither 'greedy' nor 'greedy-feedback'? Add a default setting for strat?

File: flaml/default/portfolio.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+import sys
+from sklearn.preprocessing import RobustScaler
+from flaml.default import greedy
+from flaml.default.regret import load_result, build_regret
+
+regret_bound = 0.01
+
+
+def config_predictor_tuple(tasks, configs, meta_features, regret_matrix):
+    """Config predictor represented in tuple."""
```
Comment (qingyun-wu) [inline flaml/default/portfolio.py:15]: Explain in docstr what are the elements in the tuple?

File: flaml/default/portfolio.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+    # break ties using the order in configs
+    regret = (
+        regret_matrix[tasks]
+        .loc[configs]
+        .reset_index(drop=True)
+        .apply(lambda row: row.apply(lambda x: (x, row.name)), axis=1)
+    )
+    print(regret)
+    preferences = np.argsort(regret, axis=0)
```
Comment (qingyun-wu) [inline flaml/default/portfolio.py:37]: Minor: Is `priorities` is a better name than `preferences`?

File: flaml/default/greedy.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+from sklearn.preprocessing import RobustScaler
+from sklearn.metrics import pairwise_distances
+
+
+def _augment(row):
+    max, avg, id = row.max(), row.mean(), row.index[0]
+    return row.apply(lambda x: (x, max, avg, id))
+
+
+def smart(regret_matrix, meta_features, regret_bound):
```
Comment (qingyun-wu) [inline flaml/default/greedy.py:12]: Add docstr?

File: flaml/default/suggest.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+    configs = suggest_config(task, X, y, estimator_or_predictor, location)
+    if not estimator_list:
+        return configs[0]["class"]
+    for c in configs:
+        if c["class"] in estimator_list:
+            return c["class"]
+    return estimator_list[0]
+
+
+def suggest_hyperparams(task, X, y, estimator_or_predictor, location=None):
```
Comment (qingyun-wu) [inline flaml/default/suggest.py:83]: The two function names `suggest_hyperparams` and `suggest_config` are confusing

**Review by qingyun-wu — APPROVED** [review]
